docs/tutorial/durham_reframe_tutorial.md

```diff
@@ -0,0 +1,409 @@
+
+```python
+     @sanity_function
+     def validate_solution(self):
+        return sn.assert_found(r'Solution Validates', self.stdout)
+```
+
+----
+
+## Add Performance Pattern Check
+
+To record the performance of the benchmark, ReFrame should extract a figure of merit from the output of the test. A function decorated with the `@performance_function` decorator extracts or computes a performance metric from the test’s output.
+
+> In this example, we extract four performance variables, namely the memory bandwidth values for each of the “Copy”, “Scale”, “Add” and “Triad” sub-benchmarks of STREAM, where each of the performance functions use the [`extractsingle()`](https://reframe-hpc.readthedocs.io/en/latest/deferrable_functions_reference.html#reframe.utility.sanity.extractsingle) utility function. For each of the sub-benchmarks we extract the “Best Rate MB/s” column of the output (see below) and we convert that to a float.
+
+----
+
+## Performance Pattern Check
+
+```python
+@performance_function('MB/s', perf_key='Copy')
+def extract_copy_perf(self):
+    return sn.extractsingle(r'Copy:\s+(\S+)\s+.*', self.stdout, 1, float)
+
+@performance_function('MB/s', perf_key='Scale')
+def extract_scale_perf(self):
+    return sn.extractsingle(r'Scale:\s+(\S+)\s+.*', self.stdout, 1, float)
+
+@performance_function('MB/s', perf_key='Add')
+def extract_add_perf(self):
+    return sn.extractsingle(r'Add:\s+(\S+)\s+.*', self.stdout, 1, float)
+
+@performance_function('MB/s', perf_key='Triad')
+def extract_triad_perf(self):
+    return sn.extractsingle(r'Triad:\s+(\S+)\s+.*', self.stdout, 1, float)
+```
+
+----
+
+### Perflogs
+
+- Perflogs are output in `perflogs/<system>/<partition>`
+- By default a lot of information is printed. This can be customized in the configuration file. More on this later.
+- By default not much information about build step, has to be linked back to build environment
+- See `.reframe/reports/` or use `--report-file`
+
+----
+
+## Reference values
+
+ReFrame can automate checking that the results fall within an expected range. You can set a different reference value for each `perf_key` in the performance function. For example, set the test to fail if it falls outside of +-25% of the values obtained with the previous array size.
+
+```python
+reference = {
+    'cosma': {
+        'Copy':  (40000, -0.25, 0.25, 'MB/s'),
+        'Scale': (20000, -0.25, 0.25, 'MB/s'),
+        'Add':   (20000, -0.25, 0.25, 'MB/s'),
+        'Triad': (20000, -0.25, 0.25, 'MB/s')
+    }
+}
+```
+
+> The performance reference tuple consists of the reference value, the lower and upper thresholds expressed as fractional numbers relative to the reference value, and the unit of measurement. If any of the thresholds is not relevant, None may be used instead. Also, the units in this reference variable are entirely optional, since they were already provided through the @performance_function decorator.
```

Comment:
```suggestion
The performance reference tuple consists of the reference value, the lower and upper thresholds expressed as fractional numbers relative to the reference value, and the unit of measurement. If any of the thresholds is not relevant, `None` may be used instead. Also, the units in this reference variable are entirely optional, since they were already provided through the @performance_function decorator.
```